### PR TITLE
JDK-8261791: (sctp) handleSendFailed in SctpChannelImpl.c potential leaks

### PR DESCRIPTION
**Pull request by MBaesken**
In another bug  this question from me  was answered by  Alan Bateman :

Btw. while adjusting Java_sun_nio_ch_sctp_SctpChannelImpl_receive0 , I started to wonder what happens to the allocated memory in the same file in handleSendFailed ( if ((addressP = malloc(dataLength)) == NULL) ) in early return cases incl. the CHECK_NULL , is there some deallocation missing there too ?

Yes, the error paths in handleSendFailed should be looked at. If NewDirectByteBuffer or recvmsg fails then addressP needs to be freed. Furthermore, if the NewObject fails and bufferObj != NULL then the memory for the direct buffer will need to be freed too (as JNI NewDirectByteBuffer does not setup a cleaner).

So I added freeing of the malloced memory to handleSendFailed .
Please review !

Thanks, Matthias

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261791](https://bugs.openjdk.java.net/browse/JDK-8261791): (sctp) handleSendFailed in SctpChannelImpl.c potential leaks

### Reviewers
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2586/head:pull/2586`
`$ git checkout pull/2586`

### CODE REVIEW

File: src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c [hunk 1]
```diff
@@ -229,7 +229,10 @@ void handleSendFailed
         msg->msg_iovlen = 1;
 
         bufferObj = (*env)->NewDirectByteBuffer(env, addressP, dataLength);
-        CHECK_NULL(bufferObj);
+        if (bufferObj == NULL) {
+            free(addressP);
+            return;
+        }
```
Comment (ChrisHegarty) [inline src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c:236]: This looks fine. If NewDirectByteBuffer returns NULL there will be a pending OutOfMemoryError.

File: src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c [hunk 2]
```diff
@@ -243,12 +246,14 @@ void handleSendFailed
 
         if (remaining > 0) {
             if ((rv = recvmsg(fd, msg, 0)) < 0) {
+                free(addressP);
                 handleSocketError(env, errno);
                 return;
```
Comment (ChrisHegarty) [inline src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c:251]: At this point the direct ByteBuffer has been successfully allocated and refers to the memory at addressP. But the direct ByteBuffer, while in the Java heap, will not have any references to it, so freeing addressP should be fine here.

File: src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c [hunk 3]
```diff
@@ -243,12 +246,14 @@ void handleSendFailed
...
             }
 
             if (rv != (dataLength - alreadyRead) || !(msg->msg_flags & MSG_EOR)) {
                 //TODO: assert false: "should not reach here";
+                free(addressP);
                 return;
```
Comment (ChrisHegarty) [inline src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c:257]: same a previous comment - looks ok.

File: src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c [hunk 4]
```diff
@@ -258,7 +263,10 @@ void handleSendFailed
     /* create SendFailed */
     resultObj = (*env)->NewObject(env, ssf_class, ssf_ctrID, ssf->ssf_assoc_id,
             isaObj, bufferObj, ssf->ssf_error, sri->sinfo_stream);
-    CHECK_NULL(resultObj);
+    if (resultObj == NULL) {
+        if (bufferObj != NULL) free(addressP);
+        return;
+    }
     (*env)->SetObjectField(env, resultContainerObj, src_valueID, resultObj);
```
Comment (ChrisHegarty) [inline src/jdk.sctp/unix/native/libsctp/SctpChannelImpl.c:270]: If NewObject returns NULL, there will be a pending OutOfMemoryError. Freeing addressP should be safe here, same reasoning as above.

**Review by ChrisHegarty — APPROVED** [review]

**Comment by ChrisHegarty** [comment]
This changes look good to me.  Additionally, and separately, I have filed JDK-8261881.

From 8261881 : The SCTP implementation uses NewDirectByteBuffer to create a new direct byte-buffer in native code. Such creations of a direct byte-buffer are not automatically associated with a Cleaner, therefore the native memory associated with the buffer may not be released when the buffer is no longer reachable. 

The Java code that results in the native NewDirectByteBuffer should be examined with a view to creating a cleaner and assigning it to the direct byte-buffer.

**Review by AlanBateman — APPROVED** [review]
The changes looks okay to me. I see Chris has created JDK-8261881 to setup the cleanup.

**Comment by MBaesken** [comment]
Hi Alan and Chris, thanks for the reviews.

**Comment by MBaesken** [comment]
Looks like I cannot integrate it, even after changing the title!?
Is this a bug?

**Comment by AlanBateman** [comment]
Here's the jcheck failure:
https://github.com/openjdk/jdk/pull/2586/checks?check_run_id=1918880270

**Comment by dfuch** [comment]
As pointed out by Alan: It seems like a space is missing after the colon in the PR title

**Comment by MBaesken** [comment]
> Here's the jcheck failure:
> https://github.com/openjdk/jdk/pull/2586/checks?check_run_id=1918880270

Hi Alan , I am aware of the jcheck message.
It claims : "The commit message does not reference any issue. "
I adjusted now the for the 3 or 4th time the  message  "JDK-8261791:(sctp) handleSendFailed in SctpChannelImpl.c potential leaks"  without any progress .
I think this is because  it was modified in JBS after I created it  (never had this problem without such a JBS modification).

**Comment by MBaesken** [comment]
> As pointed out by Alan: It seems like a space is missing after the colon in the PR title

lets give it a try. Thanks !

**Pull request closed without merging**